File: Corte 2/Taller 1/Insercion.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define N 1000

void insertionSort(int arr[], int n, int *comparisons, int *swaps) {
    int i, key, j;
    for (i = 1; i < n; i++) {
        key = arr[i];
        j = i - 1;

        // Comparar e insertar
        while (j >= 0 && arr[j] > key) {
            (*comparisons)++;
            arr[j + 1] = arr[j];
            j = j - 1;
            (*swaps)++;
        }
        arr[j + 1] = key;
        (*swaps)++;
    }
}

int main() {
    int arr[N];
    int comparisons = 0, swaps = 0;

    // Generar números aleatorios de 0 a 9
    srand(time(0));
    for (int i = 0; i < N; i++) {
        arr[i] = rand() % 10; // Números entre 0 y 9
    }

    // Mostrar números generados
    printf("Numeros generados:\n");
    for (int i = 0; i < N; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");

    // Ordenar usando el algoritmo de inserción
    insertionSort(arr, N, &comparisons, &swaps);

    // Mostrar números ordenados
    printf("Numeros ordenados:\n");
    for (int i = 0; i < N; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");

    // Mostrar resultados
    printf("Comparaciones: %d\n", comparisons);
    printf("Intercambios: %d\n", swaps);
    printf("Estable: Sí\n");
    printf("Insercion: Sí\n");

    return 0;
}

File: Corte 2/Taller 1/Algoritmo de Ordenamiento Go/Insercion.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

const N = 1000

func insertionSort(arr []int, comparisons *int, swaps *int) {
	for i := 1; i < N; i++ {
		key := arr[i]
		j := i - 1

		// Comparar e insertar
		for j >= 0 && arr[j] < key {
			(*comparisons)++
			arr[j+1] = arr[j]
			j--
			(*swaps)++
		}
		arr[j+1] = key
		(*swaps)++
	}
}

func main() {
	// Inicializar la semilla para números aleatorios
	rand.Seed(time.Now().UnixNano())
	// Medir el tiempo de ejecución en nanosegundos
	start := time.Now()

	arr := make([]int, N)
	for i := range arr {
		arr[i] = rand.Intn(2501) + 300000 // Números entre 300000 y 302500
	}

	// Mostrar números generados
	fmt.Println("Números generados:")
	for _, num := range arr {
		fmt.Printf("%d ", num)
	}
	fmt.Println()

	var comparisons int = 0
	var swaps int = 0

	// Ordenar usando el algoritmo de inserción
	insertionSort(arr, &comparisons, &swaps)

	// Calcular la duración
	duration := time.Since(start)

	// Mostrar números ordenados
	fmt.Println("Números ordenados:")
	for _, num := range arr {
		fmt.Printf("%d ", num)
	}
	fmt.Println()

	fmt.Printf("Comparaciones: %d\n", comparisons)
	fmt.Printf("Intercambios: %d\n", swaps)
	fmt.Println("Estable: Sí")
	fmt.Println("Inserción: Sí")

	// Mostrar el tiempo de ejecución en nanosegundos
	fmt.Printf("Tiempo de ejecución en nanosegundos: %d\n", duration.Nanoseconds())
}
